### DISCUSSION TITLE
Yolo - Black Bottle Cap Object Detection

### DISCUSSION THREAD

**Discussion by kundetiaishwarya** [ultralytics/ultralytics#22484] [Q&A]
Im training a Black shiny Cap Dataset (only caps not bottle). The cap belongs to a makeup product bottle. I want Cap top view (smooth, reflective, shiny closed side) to be class 1 and bottom view- open side to be class 2.

From bottom inside the cap, there is a **transpaent plastic semi- sphere.** This plastic semi sphere causing a light glares, reflections as well as strong refractions onto the top surface of cap from sindie.

In live camera, now bottom view is detected as class 2 (correct), Top class is also detected as class 2 (not 1). How to make the top class different from bottom class?

From images dataset i can see, top and bottom almost look same with lot of reflections.

Can anyone please help!!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kundetiaishwarya, thanks for starting a discussion and for your interest in Ultralytics 🚀! This is an automated response to help you get started. An Ultralytics engineer will also review your post and assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (e.g., distinguishing top vs. bottom views of a cap under reflections), please share:
-1–3 representative images per class showing typical challenges (reflections, glare) with annotations
-Your dataset structure, class names order, and a few sample labels
-The exact training/inference commands and full logs
-Validation metrics and any confusion matrix or per-class results
Also, please verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). These details will help us quickly diagnose labeling, data balance, or augmentation issues that can affect class separation 👍

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Thanks for the clear context—this is a classic glossy, same-object/different-view problem. First, make the views more separable in data: collect more true top-view samples under your deployment optics, balance classes, and train with milder augmentation so subtle cues aren’t washed out (e.g., try `mosaic=0.0`, reduce `hsv_*`, and keep perspective/rotation small); see the YOLO augmentation options in the [YOLO augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/) and note that if Albumentations is installed YOLO will auto-apply low‑probability blur/grayscale/CLAHE which helps with glare robustness as described in the [Albumentations integration](https://docs.ultralytics.com/integrations/albumentations/). If a single detector still confuses the classes, use a two‑stage approach: detect the cap once, crop each box, then run a tiny YOLO11 classifier trained on top vs bottom—this usually separates viewpoint cues better than a single detection head.

```python
from ultralytics import YOLO

det = YOLO('your-cap-detector.pt')      # YOLO11 detector trained on 'cap'
cls = YOLO('yolo11n-cls.pt')            # fine-tune this on {top, bottom} per the classification docs

for r in det(source=0, stream=True, imgsz=640):
    im = r.orig_img
    for xyxy in r.boxes.xyxy.int().tolist():
        x1, y1, x2, y2 = xyxy
        crop = im[y1:y2, x1:x2]
        cr = cls.predict(source=crop, imgsz=224, verbose=False)[0]
        view = cr.names[int(cr.probs.top1)]
        print('view:', view)
```

Train the classifier with folder-structured data: `yolo classify train model=yolo11n-cls.pt data=/path/to/top_bottom imgsz=224 epochs=50`; setup details are in the [classification docs](https://docs.ultralytics.com/tasks/classify/). If confusion persists, please attach a validation confusion matrix and a few misclassified top-view examples so we can suggest tighter adjustments.

**Reply by kundetiaishwarya** [reply]
Hi @glenn-jocher , thank you very much for a quick reply. I will definitely update regarding the bottle cap. 
One more question, i always have very reflective objects: for example kitkat chocolate bars with the outer covers very reflective. How to train reflective objects very well? Even with high reflection, how to get the classes correct?

By the way, im modelling object in Blender and generating synthetic dataset for objects in Blender and then train with YOLO!

**Reply by glenn-jocher** [reply]
Great follow-up—shiny packaging is mostly a lighting/domain-gap problem: randomize Blender renders heavily (swap HDRIs, move lights, vary intensity/color, tweak material roughness/clearcoat/IOR, add subtle scuffs/smudges, change camera pitch/roll), then fine-tune YOLO11 on a small set of real frames from your deployment camera; for training, bias toward photometric augments and a slightly larger `imgsz`:

```bash
yolo detect train model=yolo11s.pt data=your.yaml imgsz=768 epochs=100 \
hsv_s=0.35 hsv_v=0.45 mixup=0.05 close_mosaic=10 degrees=5 translate=0.05 scale=0.10 perspective=0.0
```

For ideas on lighting and color jitter strategies see the [Data Augmentation guide for 2025](https://www.ultralytics.com/blog/the-ultimate-guide-to-data-augmentation-in-2025) and for glare-robust transforms like CLAHE/blur used during training check the [Albumentations + YOLO article](https://www.ultralytics.com/blog/using-albumentations-augmentations-to-diversify-your-data); if you control capture, a simple lens polarizer or light diffusion also helps reduce specular hotspots.
